usr.sbin/inetd/parse_v2.c

```diff
@@ -0,0 +1,1110 @@
+		&& strcmp("rpc/udp", sep->se_proto) != 0) {
+		return true;
+	}
+	
+	if (inet_pton(AF_INET, sep->se_hostaddr, &tmp)) {
+		sep->se_family = AF_INET;
+		return true;
+	}
+
+	if (inet_pton(AF_INET6, sep->se_hostaddr, &tmp)) {
+		sep->se_family = AF_INET6;
+		return true;
+	}
+
+	ERR("Address family of %s is ambigous or invalid. "
+		"Explicitly specify protocol", sep->se_hostaddr);
+	return false;
+}
+
+/* 
+ * Skips whitespaces, newline characters, and comments, 
+ * and returns the next token. Returns false and logs error if an EOF is 
+ * encountered. 
+ */
+static bool 
+skip_whitespace(char **cpp) {
+	char *cp = *cpp;
+
+	int line_start = line_number;
+	
+	for (;;) {
+		while (isblank((unsigned char)*cp))
+			cp++;
+
+		if (*cp == '\0' || *cp == '#') {  
+			cp = nextline(fconfig);
+
+			/* Should never expect EOF when skipping whitespace */
+			if (cp == NULL) {
+				ERR("Early end of file after line %d", line_start); 
+				return false;
+			}
+			continue;
+		}
+		break;
+	}
+
+	*cpp = cp;
+	return true;
+}
+
+/* Get the key handler function pointer for the given name */
+static key_handler_func 
+get_handler(char *name) {
+	/* Call function to handle option parsing. */
+	for(size_t i = 0; i < __arraycount(key_handlers); i++) {
+		if (strcmp(key_handlers[i].name, name) == 0) {
+			return key_handlers[i].handler;
+		}
+	}
+	return NULL;
+}
+
+/* 
+ * Perform an in-place parse of a single-line quoted string 
+ * with escape sequences. Sets *cpp to the position after the quoted characters.
+ * Uses shell-style quote parsing.
+ */
+static bool 
+parse_quotes(char **cpp) {
+
+	char *cp = *cpp;
+	char quote = *cp;
+
+	memmove(cp, cp+1, strlen(cp+1)+1); 	
+	
+	while (*cp && quote) {
+		if (*cp == quote) {
+			quote = '\0';
+			memmove(cp, cp+1, strlen(cp+1)+1);
+			continue;
+		}
+		
+		if (*cp == '\\') {
+			/* start is location of backslash */
+			char *start = cp;
+			cp++;
+			switch (*cp) {
+			case 'x': {
+				char temp, bits;
+				if (((bits = hex_to_bits(*(cp + 1))) == -1) 
+					|| ((temp = hex_to_bits(*(cp + 2))) == -1)) {
+					ERR("Invalid hexcode sequence '%.4s'", start);
+					return false;
+				}
+				bits <<= 4;
+				bits |= temp;
+				*start = bits;
+				memmove(cp, cp + 3, strlen(cp+3)+1);
+				continue;
+			}
+			case '\\':
+				*start = '\\';
+				break;
+			case 'n':
+				*start = '\n';
+				break;
+			case 't':
+				*start = '\t';
+				break;
+			case 'r':
+				*start = '\r';
+				break;
+			case '\'':
+				*start = '\'';
+				break;
+			case '"':
+				*start = '"';
+				break;
+			case '\0':
+				ERR("Dangling escape sequence backslash");
+				return false;
+			default:
+				ERR("Unknown escape sequence '\\%c'", *cp);
+				return false;
+			}
+			memmove(cp, cp+1, strlen(cp+1)+1);
+			continue;
+		}
+
+		/* Regular character, advance to the next one. */
+		cp++;
+	}
+
+	if (*cp == '\0' && quote) {
+		ERR("Unclosed quote");
+		return false;
+	}
+	*cpp = cp;
+	return true;
+}
+
+char hex_to_bits(char in) {
+	switch(in) {
```

Comment:
```
unsigned char c = (unsigned char)in;
if (!isxdigit(d))
      return -1;
// ascii dependency
return isdigit(d) ? d - '0' : tolower(d) - 'a' + 10;
```
or
```
static const char xdigits[] = "0123456789abcdef";
char *p = strchr(xdigits, tolower(c));
return p == NULL ? -1 : p - xdigits;
```
or
```
        switch (in) {
        case '0'...'9':
                return in - '0';
        case 'a'...'f':
                return in - 'a' + 10;
        case 'A'...'F':
                return in - 'A' + 10;
        default:
                return -1;
        }
```

Reply:
I saw that case range operator when I was doing some google searching and wasn't sure if we were able to use that GCC feature. I'm obsessed with switch statements so I might just go with the third one. Concise and efficient (although I suppose you never know how the jump table might turn out, could just be a bunch of conditionals).

Reply:
I thought it would have made it in the c standard by now. anyway, both clang and gcc support it... and gcc does not even turn it off with gcc -ansi -std=c89, weird.